### PR TITLE
Fix DataprocSubmitJobOperator in deferrable mode=True when task is marked as failed.

### PR DESCRIPTION
**Pull request by sunank200**
When a user cancels a task for `DataprocSubmitJobOperator` in deferrable mode, the job should be cancelled in the same way as non-deferrable mode. This PR intends to fix this behaviour.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/triggers/dataproc.py [hunk 1]
```diff
@@ -91,20 +93,28 @@ def serialize(self):
...
+                state = job.status.state
+                self.log.info("Dataproc job: %s is in state: %s", self.job_id, state)
+                if state in (JobStatus.State.DONE, JobStatus.State.CANCELLED, JobStatus.State.ERROR):
+                    break
+                await asyncio.sleep(self.polling_interval_seconds)
+            yield TriggerEvent({"job_id": self.job_id, "job_state": state, "job": job})
+        except asyncio.CancelledError:
+            self.log.info("Task got cancelled.")
+            if self.job_id and self.cancel_on_kill:
+                await self.get_async_hook().cancel_job(
```
Comment (pankajkoti) [inline airflow/providers/google/cloud/triggers/dataproc.py]: Can we add a test that this method gets called by raising a CancelledError with mock.

File: airflow/providers/google/cloud/triggers/dataproc.py [hunk 2]
```diff
@@ -91,20 +93,28 @@ def serialize(self):
...
+                    job_id=self.job_id, project_id=self.project_id, region=self.region
+                )
```
Comment (pankajkoti) [inline airflow/providers/google/cloud/triggers/dataproc.py]: Were we able to test this with a DAG that the submitted job gets cancelled in GCP?

**Review by pankajkoti — COMMENTED** [review]
LGTM mostly. Minor suggestions inline.

File: airflow/providers/google/cloud/triggers/dataproc.py (hunk 1 above)
Comment (sunank200) [reply airflow/providers/google/cloud/triggers/dataproc.py:115]: Yes. I tested this

File: airflow/providers/google/cloud/triggers/dataproc.py (hunk 2 above)
Comment (sunank200) [reply airflow/providers/google/cloud/triggers/dataproc.py]: Yes, I have tested this.

File: airflow/providers/google/cloud/triggers/dataproc.py (hunk 1 above)
Comment (sunank200) [reply airflow/providers/google/cloud/triggers/dataproc.py:115]: Yes i tested this

Comment (sunank200) [reply airflow/providers/google/cloud/triggers/dataproc.py]: I have added the test

**Review by pankajkoti — APPROVED** [review]

File: airflow/providers/google/cloud/triggers/dataproc.py [hunk 3]
```diff
@@ -91,20 +93,28 @@ def serialize(self):
...
+                state = job.status.state
+                self.log.info("Dataproc job: %s is in state: %s", self.job_id, state)
+                if state in (JobStatus.State.DONE, JobStatus.State.CANCELLED, JobStatus.State.ERROR):
+                    break
+                await asyncio.sleep(self.polling_interval_seconds)
+            yield TriggerEvent({"job_id": self.job_id, "job_state": state, "job": job})
+        except asyncio.CancelledError:
+            self.log.info("Task got cancelled.")
+            if self.job_id and self.cancel_on_kill:
```
Comment (pankajastro) [inline airflow/providers/google/cloud/triggers/dataproc.py]: if `self.cancel_on_kill ` is false then shall we raise the error?

Comment (sunank200) [reply airflow/providers/google/cloud/triggers/dataproc.py]: Assuming that the user sets `self.cancel_on_kill` as False, raising an error would not be an expected behaviour. Even the non-deferrable code has the same behaviour.

File: airflow/providers/google/cloud/triggers/dataproc.py [hunk 4]
```diff
@@ -91,20 +93,28 @@ def serialize(self):
...
+                job = await self.get_async_hook().get_job(
+                    project_id=self.project_id, region=self.region, job_id=self.job_id
+                )
+                state = job.status.state
+                self.log.info("Dataproc job: %s is in state: %s", self.job_id, state)
+                if state in (JobStatus.State.DONE, JobStatus.State.CANCELLED, JobStatus.State.ERROR):
+                    break
+                await asyncio.sleep(self.polling_interval_seconds)
+            yield TriggerEvent({"job_id": self.job_id, "job_state": state, "job": job})
+        except asyncio.CancelledError:
```
Comment (Lee-W) [inline airflow/providers/google/cloud/triggers/dataproc.py:125]: Do we need to yield `TriggerEvent` or rereaise this exception?

File: tests/providers/google/cloud/triggers/test_dataproc.py [hunk 5]
```diff
@@ -375,3 +388,90 @@ async def test_async_operation_triggers_on_error(self, mock_hook, operation_trig
...
+        event = await async_gen.asend(None)
+        expected_event = TriggerEvent(
+            {"job_id": TEST_JOB_ID, "job_state": JobStatus.State.ERROR, "job": mock_hook.get_job.return_value}
+        )
+        assert event.payload == expected_event.payload
+
+    import asyncio
+    from unittest.mock import AsyncMock, patch
+
+    import pytest
```
Comment (Lee-W) [inline tests/providers/google/cloud/triggers/test_dataproc.py]: Why do we want to import them here?

Comment (sunank200) [reply tests/providers/google/cloud/triggers/test_dataproc.py]: Thanks @Lee-W I missed them why debugging

File: airflow/providers/google/cloud/triggers/dataproc.py (hunk 4 above)
Comment (sunank200) [reply airflow/providers/google/cloud/triggers/dataproc.py:125]: Yes i have added try except as well for cancellation failure now

File: airflow/providers/google/cloud/triggers/dataproc.py (hunk 3 above)
Comment (sunank200) [reply airflow/providers/google/cloud/triggers/dataproc.py]: I have raised the error now when cancellation fails.

**Review by vatsrahul1001 — APPROVED** [review]

File: airflow/providers/google/cloud/triggers/dataproc.py [hunk 6]
```diff
@@ -91,20 +104,41 @@ def serialize(self):
...
+                    self.get_sync_hook().cancel_job(
+                        job_id=self.job_id, project_id=self.project_id, region=self.region
+                    )
+                    self.log.info("Job: %s is cancelled", self.job_id)
+                    yield TriggerEvent({"job_id": self.job_id, "job_state": ClusterStatus.State.DELETING})
+            except Exception as e:
+                self.log.error("Failed to cancel the job: %s with error : %s", self.job_id, str(e))
+                raise AirflowException(
+                    f"Failed to cancel the job: {self.job_id} with error : {str(e)}"
+                ) from e
```
Comment (Lee-W) [inline airflow/providers/google/cloud/triggers/dataproc.py]: Not sure whether we really need to wrap it with `AirflowExecption`. I think we're good to just raise the original one

**Review by Lee-W — COMMENTED** [review]
Overall, it looks good to me. Left one minor discussion.

File: airflow/providers/google/cloud/triggers/dataproc.py (hunk 6 above)
Comment (sunank200) [reply airflow/providers/google/cloud/triggers/dataproc.py]: Added the change

**Comment by pankajkoti** [comment]
We will need to improvise here based on https://github.com/apache/airflow/issues/36090#issuecomment-2094972855

**Pull request merged**